Animation Editor

mattigruener opened the pull request:
Hey @johnhaddon (I hope you see this),

first version :tada: It's **very** early days, but it would be great if you follow along a bit and steer me back on the right path occasionally. I'm creating this PR in my own fork as we had said at some point that WIP PR's should be discussed outside of the main repository - and this one definitely qualifies as WIP and will for a while, I guess. Here's another screenshot that doesn't differ that much from the last one I sent:

![screenshot_20171226_084739](https://user-images.githubusercontent.com/19398650/34361028-6e865f3a-ea1b-11e7-890a-60e5cdd1720f.png)

I currently have only a few bits of the viewing capabilities and no editing at all yet:
- AnimationGadget
  Takes care of rendering the grid in the background, the axes with sensible labels, an indication for the current frame, - and holds all CurveGadgets as children.
- CurveGadget
  Holds a pointer to a `CurvePlug` and renders the data. I'm thinking that this one will eventually also be responsible for editing the data, but it might have to be controlled by the `AnimationGadget` a little in cases where changes are applied to multiple curves at the same time.
- AnimationEditor
  Just wrapping the AnimationGadget and a Widget that still needs to be written and that is meant to show the curve's names and allows operations on curves like selection/locking/hiding...

I guess the three things I'd like to tackle next are:
- Better update behaviour. 
  - When changing the size of the editor, it takes a zooming operation for the grid to be recomputed. I didn't find something like a `sizeChangedSignal`, but maybe I just missed it. I guess so far most editors didn't precompute data like that and rather did that on the fly?
  - When adding a keyframe, currently the AnimationEditor doesn't update.
- Start working on the widget to the left of the AnimationGadget. 
    I'm thinking it should be a tree view with animated plugs at least grouped by nodes, potentially even by parent plugs (`translate` and `rotate` would be examples). Plugs that control x, y, z and r, g, b components should be coloured accordingly. I am assuming that this will be a completely new Widget? I am not sure if the `PathListingWidget` can be customized enough to add things like hiding/locking curves via little icons eventually? I thought I'd ask you what you think would be the best approach.
- The `ViewportGadget` needs a way to allow non-uniform zooming. 
   While that doesn't make sense for anything else, being able to control the scale on the two axes separately is definitely needed for the `AnimationGadget`. I took a quick look at the `ViewportGadget`, but thought it'd make sense to have a quick conversation about whether we want to hack it into the `CameraController` matrix or if you have a better idea.

I'll probably continue playing with things until I you have time to have a look - will update here if I have something new. 
The first two commits are stolen from the AuxiliaryConnections PR I was working on at IE this month.

I think that's it for now. Some parts of the code are a little messy still and there are todo tags that I will have to address, but I think the general direction is worth looking at and discussing. :)


Cheerio!

Matti.

mattigruener commented:
Added three commits that:

- change the label format from 1.0000 to 1.00 (I initially thought about rendering labels for subframes, but that's not supported at the moment)
- highlight curves when hovering over them
- render a selection box

Not very useful really, but a few small things that were easy to do and move things along a little bit.

johnhaddon commented:
Matti! Sorry for the delayed response - all notifications go to my work email so I only stumbled on this just now, when I decided to have a nosey around GitHub to see if anyone had interesting branches. On my way to bed now so this'll be brief...

> CurveGadget
> Holds a pointer to a CurvePlug and renders the data. I'm thinking that this one will eventually also be responsible for editing the data, but it might have to be controlled by the AnimationGadget a little in cases where changes are applied to multiple curves at the same time.

Yeah, the division of responsibility seems a bit tricky here. I presume the current selection could span multiple curves, and a drag could therefore edit multiple curves too? At the very least I'm tempted to suggest that the CurveGadget should be totally private, so we have freedom to switch to a CurveGadgetless implementation in the future if it turns out to be an obstacle rather than a useful abstraction...

> Better update behaviour.
>When changing the size of the editor, it takes a zooming operation for the grid to be recomputed. I didn't find something like a sizeChangedSignal, but maybe I just missed it.

There's `ViewportGadget::viewportChangedSignal()` - perhaps that is what you need?

> I guess so far most editors didn't precompute data like that and rather did that on the fly?

Yep, I think so. Is the precomputation essential here for performance then?

> When adding a keyframe, currently the AnimationEditor doesn't update.

I think the only option here is to use `plugDirtiedSignal()`, which is triggered for the output of the curve plug.

> Start working on the widget to the left of the AnimationGadget.
I'm thinking it should be a tree view with animated plugs at least grouped by nodes, potentially even by parent plugs (translate and rotate would be examples).

Sounds good.

> I am assuming that this will be a completely new Widget? I am not sure if the PathListingWidget can be customized enough to add things like hiding/locking curves via little icons eventually? I thought I'd ask you what you think would be the best approach.

This is a good question. It's tempting to write something new and custom-tailored using Qt directly, but there's also a good argument for "eating your own dog food", and using the PathListingWidget. It always annoyed me that a certain other software package had a much more limited set of widgets for use via scripting than they used internally, so I do try as much as possible to build all Gaffer's UIs using GafferUI widgets rather than Qt widgets. We will need to add features to the PathListingWidget to support more stuff in the SceneHierarchy anyway, so if it's not crazy urgent to add the hiding/locking, I think perhaps it's worth starting out with a PathListingWidget to see how far we can get.

> The ViewportGadget needs a way to allow non-uniform zooming.
While that doesn't make sense for anything else, being able to control the scale on the two axes separately is definitely needed for the AnimationGadget. I took a quick look at the ViewportGadget, but thought it'd make sense to have a quick conversation about whether we want to hack it into the CameraController matrix or if you have a better idea.

I think I'm right in saying that for orthographic views, the "zooming" is actually implemented by manipulating the screen window rather than doing anything with the camera matrix. If so, we should do the same to support the non-uniform scaling.

mattigruener commented:
Hey John,

bit of an update here. I won't get to do much more on the weekend so I'm putting up the latest code that produces something like in the screenshot below.

![screenshot_20180114_100143](https://user-images.githubusercontent.com/19398650/34919030-406bdb2e-f912-11e7-8633-84f6509c47be.png)

New things:
- Widget on the left that controls which curves are visible similar to what the `SceneHierarchy` is doing.
- Non-uniform dollying.
- Color-coding of curves based on name
- Basic selection functionality
- Computing the grid isn't cached anymore like before, but it's done when rendering.

There's stuff I need to look at again and that needs some cleaning up, because I've been shuffling things around a bit - so maybe we take this more as a starting point for high-level discussions and less for looking at the code in detail..? I've squashed all the new stuff into the old commits where possible. 

Two bits that would definitely benefit from an additional pair of eyes (your's in particular) already are 
- what I've done in `ViewportGadget` to enable non-uniform dollying.
- that mini-fix I've added in `GraphComponentPath`

If anything else jumps at you as the wrong way of doing things, let me know of course :)

Things on my list of things to fix (apart from a bunch of todos in the code):
- `CurveGadget` needs to be more private
- Keys need to always render above curves
- Use `plugDirtiedSignal` for updates
- Needs tests.

In terms of UX, I think it's a bit unfortunate that it's not possible to view curves in isolation if the respective plugs have the same parent. But I guess we've lived with that limitation in the SceneHierarchy as well and it hasn't come up as a big problem.

Maybe the next step, after addressing some of the todos, is to insert a key at the current frame into all visible curves and start moving selected keys around? About that last one, did you want to just attach the selected keys to the mouse when middle-dragging for example or should we add modulators like you've done in the SceneView? 
I started doing 3D stuff years ago in an early Blender and always liked how hitting a hotkey just attaches the selection to the mouse with an option to hit x|y|z to restrict movement to a particular axis. Probably not a good idea to introduce very different concepts into Gaffer at this point.

And we should maybe have a quick chat about what we consider necessary as a minimal set of features to get this PR merged? Is inserting/deleting and moving keys enough to get something useful into peoples' hands so we can get some early feedback?

Cheerio for now :)

Matti.

johnhaddon commented:
Great to see further progress Matti!

> There's stuff I need to look at again and that needs some cleaning up, because I've been shuffling things around a bit - so maybe we take this more as a starting point for high-level discussions and less for looking at the code in detail..?

I have refrained almost entirely from looking at the code :)

> two bits that would definitely benefit from an additional pair of eyes (your's in particular) already are :
> what I've done in ViewportGadget to enable non-uniform dollying.

Something along these lines seems totally reasonable. The "nonUniformDolly" naming seems a little odd - dollying is moving the camera, which only happens when we're using a perspective camera. We've overloaded the name for use with orthographic cameras (where we actually zoom), which is already a bit odd, but to then introduce a "nonUniformDolly" that doesn't apply to perspective at all is making it all the more strange. Don't suppose you have any thoughts for a better name? Or maybe we just document it better?

> that mini-fix I've added in GraphComponentPath

That looks spot-on to me. Could you include a test?

> Maybe the next step, after addressing some of the todos, is to insert a key at the current frame into all visible curves and start moving selected keys around? About that last one, did you want to just attach the selected keys to the mouse when middle-dragging for example or should we add modulators like you've done in the SceneView?

I presume you mean manipulators rather than modulators? I think in the long run we'll want to introduce different manipulators for translate and scale. But I would expect the default manipulator to be select/translate, so I think it's fine to just hardcode the interaction for that now, and then refactor into different modes/manipulators later on.

> I started doing 3D stuff years ago in an early Blender and always liked how hitting a hotkey just attaches the selection to the mouse with an option to hit x|y|z to restrict movement to a particular axis. Probably not a good idea to introduce very different concepts into Gaffer at this point.

I'd like to keep the click/drag approach to initiating drags rather than initiate a drag with a keypress. But we definitely need a way of restricting movement to a particular axis, and this should be in the first version I think. From Inkscape I'm used to holding down Control to restrict motion to the primary direction of the mouse move - do you think that might work?

By default, we probably also want to snap to the nearest frame when moving in the X axis?

> And we should maybe have a quick chat about what we consider necessary as a minimal set of features to get this PR merged? Is inserting/deleting and moving keys enough to get something useful into peoples' hands so we can get some early feedback?

Yep, that sounds good to me.

mattigruener commented:
> Don't suppose you have any thoughts for a better name? Or maybe we just document it better?

Yeah, I wasn't sure what to call it. I'll ask around at work today to see if someone has a better idea.

> That looks spot-on to me. Could you include a test?

Will do!

> I presume you mean manipulators rather than modulators?

Of course -.-

> From Inkscape I'm used to holding down Control to restrict motion to the primary direction of the mouse move - do you think that might work?

Yeah, that sounds good. Maybe for both this feature and the "nonUniformDolly" I can also look into swapping the mouse cursor to something that indicates which axis is being used..? I think I've seen that in other tools and maybe it's a nice addition in terms of UX. And maybe it's not important at all at this point ;)
Not sure if you've tried the dollying - I thought it behaves a little odd when switching from one axis to the other because there's a little jump based on how far the mouse has already moved. Maybe that's ok - will check what other tools are doing.

> By default, we probably also want to snap to the nearest frame when moving in the X axis?

Yeah, I think so. I was wondering (for both axes) if snapping should happen only to the grid's lines. So that you'd have to zoom in to snap frame by frame and get snapping to every fifth, thenth, whateverth frame otherwise. Maybe that's more useful for the y-axis? The y-axis needs some work to add intervals smaller than 1, too.

johnhaddon commented:
> Yeah, I wasn't sure what to call it. I'll ask around at work today to see if someone has a better idea.

Maybe variableAspectZoom? Bit wordy, but I think it describes what is happening fairly well.

> I can also look into swapping the mouse cursor to something that indicates which axis is being used..?

Sounds like a nice touch.

> Not sure if you've tried the dollying - I thought it behaves a little odd when switching from one axis to the other because there's a little jump based on how far the mouse has already moved. Maybe that's ok - will check what other tools are doing.

I haven't tried anything yet - I'm just going off the screenshots and a quick peek at the code. That does sound like it might be a bit confusing.

> I was wondering (for both axes) if snapping should happen only to the grid's lines. So that you'd have to zoom in to snap frame by frame and get snapping to every fifth, thenth, whateverth frame otherwise.

I don't think snapping to the grid should be the default, and I wouldn't expect the Y axis to default to snapping either. I just suggested the snap-to-frame because otherwise you rarely get what you expect when it comes to render...

I am of course saying all this from a position of relative ignorance - if anyone with more of a stake in animation has other views, we should probably give them priority...

mattigruener commented:
> Maybe variableAspectZoom? Bit wordy, but I think it describes what is happening fairly well.

That sounds better, yes. 

> That does sound like it might be a bit confusing.

I will play with resetting the amount of scaling when switching to the other axis. Might be better.

> I don't think snapping to the grid should be the default, and I wouldn't expect the Y axis to default to snapping either. I just suggested the snap-to-frame because otherwise you rarely get what you expect when it comes to render...

Oh I didn't mean to suggest to default to snapping for y, sorry. I was thinking more that when an additional key is pressed you get snapping and I thought that the lines would be a good indication where those snaps would land. I agree that we should always snap to full frames for the x-axis.

mattigruener commented:
Small update - probably not worth talking about just yet. Just wanted to put up the code changes as additional backup. I still have to address some stuff we already talked about.

Changes:
- changed the way scaling individual axes works. no jumps anymore and mouse pointer changes icon
- inserting keyframes at current frame
- removing selected keyframes
- ctrl-dragging to expand current selection
- expanding and selecting in the widget on the left works as discussed in separate email thread

mattigruener commented:
Small update. I'm not happy with how the `AnimationGadget` is updating `Key`'s on `CurvePlugs`, - something we should talk about at some point. I've also put some reminders about that in the code. It's potentially the biggest question mark currently...

I think all functionality we'd expect for a first release is there now, but I have a bunch of things to clean up and I need to write tests. I think I'll tidy up some bits next and then address all the changes that you'll request once you've seen what I've done here ;) Tests afterwards.

Changes:
- Keyframes can be moved now
  - they snap to frames
  - holding shift let's users move in a single axis (mouse pointer changes accordingly)
- hitting 'f' performs a somewhat useful framing based on current selection/visibility

mattigruener commented on src/GafferUI/AnimationGadget.cpp:
```diff
@@ -0,0 +1,773 @@
+		{
+			// protect existing keys by adding small epsilon
+			// \todo store epsilon somewhere more global
+			key.time += 0.00001;
```
Note to myself: This is messing with the snapping to integer frames and potentially causes confusion...

mattigruener replied:
Just snap it to the closest frame without a key.

mattigruener commented on src/GafferUI/AnimationGadget.cpp:
```diff
@@ -434,55 +434,50 @@ void AnimationGadget::removeKeyframes()
+		if( m_moveAxis != MoveAxis::X )
 		{
-			key.value += offset.y;
+			index.modify( uniqueKeyIt, UniqueKeyChangeValue( uniqueKeyIt->first.value + offset.y ) );
```
Note to self: This is currently adding the full offset every single time and needs to be fixed.
